pkg/datarepair/datarepair_test.go

```diff
@@ -126,6 +140,139 @@ func TestDataRepair(t *testing.T) {
 		remotePieces = pointer.GetRemote().GetRemotePieces()
 		for _, piece := range remotePieces {
 			assert.False(t, nodesToKill[piece.NodeId])
+			assert.False(t, nodesToDisqualify[piece.NodeId])
+		}
+	})
+}
+
+func TestRepairMultipleDisqualified(t *testing.T) {
+	testplanet.Run(t, testplanet.Config{
+		SatelliteCount:   1,
+		StorageNodeCount: 12,
+		UplinkCount:      1,
+	}, func(t *testing.T, ctx *testcontext.Context, planet *testplanet.Planet) {
+		// first, upload some remote data
+		ul := planet.Uplinks[0]
+		satellite := planet.Satellites[0]
+		// stop discovery service so that we do not get a race condition when we delete nodes from overlay cache
+		satellite.Discovery.Service.Discovery.Stop()
+		satellite.Discovery.Service.Refresh.Stop()
+
+		satellite.Repair.Checker.Loop.Pause()
+		satellite.Repair.Repairer.Loop.Pause()
+
+		testData := make([]byte, 1*memory.MiB)
+		_, err := rand.Read(testData)
+		assert.NoError(t, err)
+
+		err = ul.UploadWithConfig(ctx, satellite, &uplink.RSConfig{
+			MinThreshold:     3,
+			RepairThreshold:  5,
+			SuccessThreshold: 7,
+			MaxThreshold:     7,
+		}, "testbucket", "test/path", testData)
+		require.NoError(t, err)
+
+		// get a remote segment from metainfo
+		metainfo := satellite.Metainfo.Service
+		listResponse, _, err := metainfo.List(ctx, "", "", "", true, 0, 0)
+		require.NoError(t, err)
+
+		var path string
+		var pointer *pb.Pointer
+		for _, v := range listResponse {
+			path = v.GetPath()
+			pointer, err = metainfo.Get(ctx, path)
+			assert.NoError(t, err)
+			if pointer.GetType() == pb.Pointer_REMOTE {
+				break
+			}
+		}
+
+		// calculate how many storagenodes to disqualify
+		numStorageNodes := len(planet.StorageNodes)
+		redundancy := pointer.GetRemote().GetRedundancy()
+		remotePieces := pointer.GetRemote().GetRemotePieces()
+		minReq := redundancy.GetMinReq()
+		numPieces := len(remotePieces)
+		toDisqualify := numPieces - (int(minReq + 1))
+		// we should have enough storage nodes to repair on
+		assert.True(t, (numStorageNodes-toDisqualify) >= numPieces)
+
+		// disqualify nodes and track lost pieces
+		var lostPieces []int32
+		nodesToDisqualify := make(map[storj.NodeID]bool)
+		nodesToKeepAlive := make(map[storj.NodeID]bool)
+
+		for i, piece := range remotePieces {
+			if i >= toDisqualify {
+				nodesToKeepAlive[piece.NodeId] = true
+				continue
+			}
+			nodesToDisqualify[piece.NodeId] = true
+			lostPieces = append(lostPieces, piece.GetPieceNum())
 		}
+
+		for _, node := range planet.StorageNodes {
+			if nodesToDisqualify[node.ID()] {
+				disqualifyNode(t, ctx, satellite, node.ID())
+			}
+		}
+
+		satellite.Repair.Checker.Loop.Restart()
+		satellite.Repair.Checker.Loop.TriggerWait()
+		satellite.Repair.Checker.Loop.Pause()
+		satellite.Repair.Repairer.Loop.Restart()
+		satellite.Repair.Repairer.Loop.TriggerWait()
+		satellite.Repair.Repairer.Loop.Pause()
+		satellite.Repair.Repairer.Limiter.Wait()
+
+		// kill nodes kept alive to ensure repair worked
```

Comment:
I'm missing something here.
If repair is executed before this nodes are killed, the only pieces to repair are the ones which were store in the disqualified nodes, so if we kill these nodes and repair isn't run again, how is it guarantee that the number of pieces which remain in the network (nodes alive an not disqualified) are the minimum for being able to retrieve the data?

Reply:
Here is what happens:
- Test data is uploaded to 7 nodes
- 3 of them are disqualified
- Data repair is triggered. It repairs the data from the remaining 4 nodes to additional 3 new nodes
- The 4 nodes from which the data was repaired are shut down.
- Now we have just the 3 new nodes to which the data was repaired.
- We try to download the data from these 3 nodes. It is successful because 3 nodes are enough for download.